File: src/main/java/io/github/devcavin/netinsight/service/IpService.java
package io.github.devcavin.netinsight.service;

import io.github.devcavin.netinsight.dto.IpInfo;
import io.github.devcavin.netinsight.enums.IpStatus;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.net.*;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.*;

@Service
public class IpService {
    private static final Logger log = LoggerFactory.getLogger(IpService.class);

    /**
     * Retrieves local IPs for all active interfaces, using single IPv4 + IPv6 per interface.
     */
    private Map<String, IpInfo.IpPair> getLocalIps() {
        Map<String, IpInfo.IpPair> result = new LinkedHashMap<>();

        try {
            Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();

            while (interfaces.hasMoreElements()) {
                NetworkInterface ni = interfaces.nextElement();
                if (!ni.isUp() || ni.isLoopback()) continue;

                String ipv4 = null;
                String ipv6 = null;

                Enumeration<InetAddress> addresses = ni.getInetAddresses();
                while (addresses.hasMoreElements()) {
                    InetAddress addr = addresses.nextElement();
                    if (addr.isLoopbackAddress()) continue;

                    String ip = addr.getHostAddress();
                    int percentIndex = ip.indexOf('%'); // remove scope for IPv6
                    if (percentIndex != -1) ip = ip.substring(0, percentIndex);

                    if (addr instanceof Inet4Address && ipv4 == null) ipv4 = ip;
                    else if (!(addr instanceof Inet4Address) && ipv6 == null) ipv6 = ip;

                    if (ipv4 != null && ipv6 != null) break;
                }

                if (ipv4 != null || ipv6 != null) {
                    result.put(ni.getName(), new IpInfo.IpPair(ipv4, ipv6));
                }
            }

        } catch (Exception e) {
            log.error("Error while retrieving local IPs", e);
            return Collections.emptyMap();
        }

        return result;
    }

    /**
     * Fetches the public IP from a given URL.
     */
    private String fetchPublicIp(String url) throws Exception {
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .GET()
                .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
        return response.body();
    }

    /**
     * Returns IP information including local and public addresses.
     * Public IPv6 will be null if unavailable from the ISP.
     */
    public IpInfo getIpInfo() {
        try {
            Map<String, IpInfo.IpPair> localIps = getLocalIps();

            // Public IPv4
            String publicIpv4 = fetchPublicIp("https://api.ipify.org");

            // Public IPv6, null if ISP doesn't allow
            String publicIpv6 = null;
            try {
                InetAddress test = InetAddress.getByName("api64.ipify.org");
                if (test instanceof Inet6Address) {
                    publicIpv6 = fetchPublicIp("https://api64.ipify.org");
                }
            } catch (Exception ignored) {
                // ISP doesn't provide IPv6
            }

            IpInfo.IpPair publicIp = new IpInfo.IpPair(publicIpv4, publicIpv6);

            return new IpInfo(localIps, publicIp, IpStatus.SUCCESS, null);

        } catch (Exception e) {
            log.error("Failed to retrieve IP info", e);
            return new IpInfo(null, null, IpStatus.ERROR, e.getMessage());
        }
    }
}

File: src/main/kotlin/io/github/devcavin/netinsight/service/IpService.kt
package io.github.devcavin.netinsight.service

import io.github.devcavin.netinsight.domain.LocalIpProvider
import io.github.devcavin.netinsight.domain.PublicIpClient
import io.github.devcavin.netinsight.dto.response.IpResponse
import io.github.devcavin.netinsight.enum.Status
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.net.Inet6Address
import java.net.InetAddress

@Service
class IpService(
    private val localIpProvider: LocalIpProvider,
    private val publicIpClient: PublicIpClient
) {

    private val log = LoggerFactory.getLogger(IpService::class.java)

    fun getIpInfo(): IpResponse {
        // Fetch local IPs
        val localIps = try {
            val ips = localIpProvider.getLocalIps()
            ips.ifEmpty { null }
        } catch (e: Exception) {
            log.warn("Failed to fetch local IPs", e)
            null
        }

        // Fetch public IPv4
        val ipv4 = try {
            publicIpClient.getIpv4()
        } catch (e: Exception) {
            log.warn("Failed to fetch public IPv4", e)
            null
        }

        // Fetch public IPv6
        val rawIpv6 = try {
            publicIpClient.getIpv6()
        } catch (e: Exception) {
            log.warn("Failed to fetch public IPv6", e)
            null
        }

        val ipv6 = if (rawIpv6 != null && isValidIpv6(rawIpv6)) {
            rawIpv6
        } else {
            "not configured by ISP"
        }

        // Determine if any IP exists
        val hasLocalIps = !localIps.isNullOrEmpty()
        val hasPublicIp = ipv4 != null || (rawIpv6 != null && isValidIpv6(rawIpv6))

        val isOffline = !hasLocalIps && !hasPublicIp

        val status = if (isOffline) Status.ERROR else Status.SUCCESS

        val message = if (isOffline) {
            "Failed to retrieve any IP information."
        } else {
            "IP information retrieved successfully."
        }

        // Return empty maps for offline, otherwise full data
        val localIpResult = if (isOffline) emptyMap() else localIps ?: emptyMap()
        val publicIpResult = if (isOffline) IpResponse.IpPair(null, null)
        else IpResponse.IpPair(ipv4, ipv6)

        return IpResponse(
            localIp = localIpResult,
            publicIp = publicIpResult,
            status = status,
            message = message
        )
    }

    private fun isValidIpv6(ip: String): Boolean {
        return try {
            InetAddress.getByName(ip) is Inet6Address
        } catch (e: Exception) {
            false
        }
    }
}